### PR TITLE
fix: DeletedObject constraint error when delete ProgramRule

### PR DESCRIPTION
**Pull request by vietnguyen**
https://jira.dhis2.org/browse/DHIS2-3454

- The issue is that for saving `DeletedObject` we use `DeletedObjectIntegrator` for triggering Hibernate POST_DELETE event, which is triggered for every objects including cascade objects
But for deleting the `DeletedObject` we explicit call `deleteDeletedObject` method in `HibernateIdentifiableObjectStore.save()` which does not touch the cascade objects.

- This fix remove `DeletedObjectIntegrator` and use Hibernate POST_COMMIT_INSERT event to call `deleteDeletedObject`, so the cascade objects will be taken care of. Also adding `DeletedObjectListenerConfigurer` to register those hibernate event listeners

- Also include a small fix for json serialization issue of `ProgramRuleAction.evaluationEnvironment` which is not an `IdentifiableObject` but was marked with `@JsonSerialize( contentAs = BaseIdentifiableObject.class )`

After this fix, there were some integration test failures. The reason is because we have those kinds of code
```
dataElementStore.save( dataElementA );
categoryService.getDefaultCategoryCombo();
```
 

- In first line, hibernate put dataElementA into persistentContext of the current session, no transaction.
- Then in second line, a new *readOnly* transaction is created in same session.
- Therefore, when the transaction is committed/session flushed, hibernate can't save dataElementA in a readOnly transaction and throws error.
- Before the fix, we call below code in `HibernateIdentifiableStore` so a separate transaction with readOnly = false is created when saving dataElementA. 
```
 if (object instanceof MetadataObject )
 {
       deletedObjectService.deleteDeletedObjects( new DeletedObjectQuery( object ) );
 }
```

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deletedobject/hibernate/DeletedObjectListenerConfigurer.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package org.hisp.dhis.deletedobject.hibernate;
+
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deletedobject/hibernate/DeletedObjectListenerConfigurer.java]: Add license.

**Comment by vietnguyen** [comment]
This is still WIP, there is an integration test error, I'm fixing it.

**Comment by stian-sandvold** [comment]
Looks good to me. Could you have a look at these issues as well @vietnguyen and see if you can at least resolve the non-comment one?

https://sonarcloud.io/project/issues?id=dhis2_dhis2-core&pullRequest=5187&resolved=false&severities=CRITICAL&sinceLeakPeriod=true

**Comment by vietnguyen** [comment]
> Looks good to me. Could you have a look at these issues as well @vietnguyen and see if you can at least resolve the non-comment one?
> 
> https://sonarcloud.io/project/issues?id=dhis2_dhis2-core&pullRequest=5187&resolved=false&severities=CRITICAL&sinceLeakPeriod=true

@stian-sandvold  Thanks, done.

**Pull request merged**